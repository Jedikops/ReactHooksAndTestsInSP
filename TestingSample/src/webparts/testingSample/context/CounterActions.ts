import ICounter from "./ICounter";

export type CounterAction = (state: ICounter, value: number) => ICounter;

export const CounterActions = {
    Increment: (state: ICounter, value: number): ICounter => {
        return { count: state.count + value };
    },

    Decrement: (state: ICounter, value: number): ICounter => {
        return { count: state.count - value };
    },

    Multiply: (state: ICounter, value: number): ICounter => {
        return { count: state.count * value };
    },

    Divide: (state: ICounter, value: number): ICounter => {
        return { count: state.count / value };
    },

    ResetTo: (state: ICounter, value: number): ICounter => {
        return { count: value };
    }
}