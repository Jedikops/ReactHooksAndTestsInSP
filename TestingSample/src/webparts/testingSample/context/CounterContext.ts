import * as React from "react";
import { createContext, useContext } from "react";
import ICounterContext from "./ICounterContext";

export const CounterContext: React.Context<ICounterContext> = createContext({} as ICounterContext);

export const useStore = (): ICounterContext => useContext(CounterContext);