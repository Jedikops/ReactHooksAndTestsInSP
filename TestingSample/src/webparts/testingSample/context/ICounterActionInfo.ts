import { CounterAction } from "./CounterActions";

export default interface ICounterActionInfo {
    action: CounterAction;
    value: number;

}