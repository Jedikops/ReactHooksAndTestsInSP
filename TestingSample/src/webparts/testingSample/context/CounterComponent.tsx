import * as React from "react";
import { useStore } from "./CounterContext";
import ICounterContext from "./ICounterContext";


export const CounterComponent = () => {
    const { state }: ICounterContext = useStore();

    return (
        <div>
            <div className="count">{state.count}</div>
        </div>
    );
};