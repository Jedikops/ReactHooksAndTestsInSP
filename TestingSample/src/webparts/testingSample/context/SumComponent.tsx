import * as React from "react";
import { useStore } from "./CounterContext";
import ICounterState from "./ICounterContext";
import { CounterActions } from "./CounterActions";


export const SumComponent = (props: { component: number}) => {
    const { state, dispatch }: ICounterState = useStore();

    React.useEffect(() => {
        dispatch({ action: CounterActions.ResetTo, value: 0 });
    }, []);

    return (
        <div>
            <div>
                <button className="firstButton" onClick={() => dispatch({ action: CounterActions.Increment, value: props.component })}>+</button>
                <button className="secondButton" onClick={() => dispatch({ action: CounterActions.Decrement, value: props.component })}>-</button>
                <button className="thirdButton" onClick={() => dispatch({ action: CounterActions.ResetTo, value: 0 })}>Reset</button>
            </div>
        </div>
    );
};