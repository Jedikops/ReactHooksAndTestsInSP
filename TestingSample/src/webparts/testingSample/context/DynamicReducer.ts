import { Reducer } from "react";
import ICounter from "./ICounter";
import ICounterActionInfo from "./ICounterActionInfo";

export const DynamicReducer: Reducer<ICounter, ICounterActionInfo> = (state, actionInfo) : ICounter=>{
    return actionInfo.action(state, actionInfo.value);
}