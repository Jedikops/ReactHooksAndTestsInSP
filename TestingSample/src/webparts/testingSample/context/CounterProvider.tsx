import * as React from 'react';
//import { CounterReducer } from "./CounterReducers";
import { CounterContext } from "./CounterContext";
import ICounter from "./ICounter";
import { useReducer, Reducer } from 'react';
import { DynamicReducer } from './DynamicReducer';
import ICounterActionInfo from './ICounterActionInfo';

const initialState: ICounter = { count: 0 };

export const CounterProvider = ({ children }) => {
    const [state, dispatch] = useReducer<Reducer<ICounter, ICounterActionInfo>>(DynamicReducer, initialState);
React.useState
    return (<CounterContext.Provider value={{ state, dispatch: dispatch }}>
        {children}
    </CounterContext.Provider >);

};

