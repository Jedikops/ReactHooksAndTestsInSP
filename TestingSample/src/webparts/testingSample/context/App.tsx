import * as React from 'react';
import { CounterProvider } from "./CounterProvider";
import { SumComponent } from "./SumComponent";
import { MultiplyComponent } from './MultiplyComponent';
import { CounterComponent } from './CounterComponent';

const App = () => {
    return (
        <CounterProvider>
            <CounterComponent />
            <SumComponent component={1} />
            <MultiplyComponent multiplier={2} />
        </CounterProvider >
    );
};
export default App;
