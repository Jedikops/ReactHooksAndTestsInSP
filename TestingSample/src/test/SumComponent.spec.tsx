import * as React from 'react';
import 'jest';
import { mount, ReactWrapper } from 'enzyme';

import { SumComponent } from '../webparts/testingSample/context/SumComponent';
import { CounterProvider } from '../webparts/testingSample/context/CounterProvider';
import { CounterComponent } from '../webparts/testingSample/context/CounterComponent';


let initialValue = 0;
let component = 1;
let expected = initialValue + component;

let wrapper: ReactWrapper;

beforeEach(() => {
    wrapper = mount(<CounterProvider>
        <CounterComponent />
        <SumComponent component={component} />
    </CounterProvider>);
});

afterEach(() => {
    wrapper.unmount();
});

test('initial test to verify inital count of counter', () => {
    expect(wrapper).toMatchSnapshot();
    //expect(wrapper.find('div.count').text()).toBe((initialValue).toString());
});

test('first button restult check', () => {
    expect(wrapper.find('button.firstButton')).not.toBeNull();
    let button = wrapper.find('button.firstButton').first();
    button.simulate('click');

    expect(wrapper).toMatchSnapshot();
    //expect(wrapper.find('div.count').text()).toBe((expected).toString());
});



