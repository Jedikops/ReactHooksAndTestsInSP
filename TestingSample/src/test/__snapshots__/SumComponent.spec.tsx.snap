// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`first button restult check 1`] = `
<Component>
  <Component>
    <div>
      <div
        className="count"
      >
        1
      </div>
    </div>
  </Component>
  <Component
    component={1}
  >
    <div>
      <div>
        <button
          className="firstButton"
          onClick={[Function]}
        >
          +
        </button>
        <button
          className="secondButton"
          onClick={[Function]}
        >
          -
        </button>
        <button
          className="thirdButton"
          onClick={[Function]}
        >
          Reset
        </button>
      </div>
    </div>
  </Component>
</Component>
`;

exports[`initial test to verify inital count of counter 1`] = `
<Component>
  <Component>
    <div>
      <div
        className="count"
      >
        0
      </div>
    </div>
  </Component>
  <Component
    component={1}
  >
    <div>
      <div>
        <button
          className="firstButton"
          onClick={[Function]}
        >
          +
        </button>
        <button
          className="secondButton"
          onClick={[Function]}
        >
          -
        </button>
        <button
          className="thirdButton"
          onClick={[Function]}
        >
          Reset
        </button>
      </div>
    </div>
  </Component>
</Component>
`;
